### PR TITLE
feat: #411 Scheduled Tasks with interval-based scheduling and per-project webhooks

### LINKED ISSUE #411: feat: Scheduled Tasks - Auto-execution of webhooks with cron-like scheduling
## Problem

Currently, webhooks can only be triggered manually via HTTP calls. For recurring tasks (daily reports, periodic scans, monitoring, etc.), users must set up external cron jobs or schedulers to call webhook endpoints.

This adds complexity and requires infrastructure outside of Coday, especially for local development environments.

## Proposed Solution

Add scheduled triggers that execute existing webhooks on a cron schedule.

### Architecture

**Triggers**: Lightweight scheduling layer that references existing webhooks
- Stored per-project: `~/.coday/projects/{project}/triggers/`
- One trigger = one cron schedule
- References a webhook for execution

**Structure**:
```yaml
id: "trigger-123"
name: "Daily Security Scan"
enabled: true
webhookUuid: "webhook-abc-123"  # Reference to existing webhook
schedule: "0 9 * * *"            # Cron expression
parameters:                      # Override webhook parameters, examples below
  severity: "high"
  focus_area: "authentication"
createdBy: "username"
createdAt: "2025-12-20T10:00:00Z"
lastRun: "2025-12-20T09:00:00Z"
nextRun: "2025-12-21T09:00:00Z"
```

### Supported Cron Patterns (MVP)

Keep it simple for local dev:
- `*/X * * * *` - Every X minutes
- `0 */X * * *` - Every X hours  
- `0 0 * * *` - Daily at midnight
- `0 0 * * 0` - Weekly on Sunday

### REST API

```
GET    /api/projects/:projectName/triggers           # List all
GET    /api/projects/:projectName/triggers/:id       # Get details
POST   /api/projects/:projectName/triggers           # Create
PUT    /api/projects/:projectName/triggers/:id       # Update
DELETE /api/projects/:projectName/triggers/:id       # Delete

POST   /api/projects/:projectName/triggers/:id/enable   # Enable
POST   /api/projects/:projectName/triggers/:id/disable  # Disable
POST   /api/projects/:projectName/triggers/:id/run-now  # Manual trigger
```

### Example Usage

1. Create a webhook template for security scanning (if not exists)
2. Create a trigger pointing to that webhook with daily schedule
3. Trigger automatically executes webhook at scheduled time
4. Results logged and thread auto-cleaned after timeout

## Implementation

**TriggerService** (similar to ThreadCleanupService pattern):
- Load all triggers from all projects at startup
- Check every minute if any trigger should execute
- Execute via existing webhook endpoint
- Update lastRun/nextRun timestamps
- Audit logging

**Files**:
- `libs/model/trigger.ts` - TypeScript interfaces
- `libs/service/trigger.service.ts` - CRUD + scheduling logic
- `apps/server/src/trigger.routes.ts` - REST API routes

## Implementation Effort

**Estimated: 8-12 hours**

- TriggerService (6-8h): CRUD + scheduling + cron parsing
- REST API routes (2-3h): Validation + error handling
- Manual testing (1h)

Why reasonable?
- ✅ Timer infrastructure exists (ThreadCleanupService pattern)
- ✅ Webhooks fully functional (just reference them)
- ✅ Logging system in place
- ✅ YAML storage pattern established
- ✅ No external dependencies needed

## Scope

**Phase 1 (MVP)**: Backend only
- TriggerService with basic cron parsing (4 patterns)
- REST API for CRUD operations
- Per-project storage (not multi-project like current webhooks)
- Audit logging

**Phase 2 (Optional)**: Web UI
- List/create/edit triggers
- Enable/disable toggles
- Execution history
- Run-now button for testing

**Future (DB Migration)**: 
- Prompt chains (reusable action definitions)
- More trigger types (event-driven, conditional, etc.)
- Webhook refactoring to per-project storage

## Limitations (Acceptable for Local Dev)

- ✅ Minute precision (no seconds) - standard for cron
- ✅ Single instance (no clustering) - local environment
- ✅ UTC timezone only - simplification
- ✅ State recalculated on restart - acceptable

## Related

- Leverages existing webhook system (no changes needed)
- Similar pattern to `ThreadCleanupService`
- Uses `ThreadCodayManager` for execution
- Integrates with `CodayLogger` for audit trails
- Per-project storage (consistent with threads)

### PR DESCRIPTION
**Pull request by vincent-audibert-whoz**
## Summary

This PR introduces a comprehensive **Prompts and Schedulers** system that replaces the previous webhook-based approach with a more flexible and user-friendly architecture for reusable AI tasks.

### 1. Prompt System
**Reusable AI task definitions** that can be executed:
- **Directly by users** through the web interface
- **Via schedulers** for automated execution
- **Via webhooks** for external API triggers (CODAY_ADMIN only)

**Key capabilities**:
- Flexible parameterization (simple string or structured key-value pairs)
- `{{PARAMETERS}}` placeholder for simple values
- `{{key}}` placeholders for structured parameters
- Auto-append mode when no placeholders present
- Strict validation with clear error messages
- Normalized naming (lowercase-with-hyphens)
- Collaborative editing (all users can edit all prompts)

### 2. Scheduler System
**Automated prompt execution** with interval-based scheduling:
- Minutes (1-59min), Hours (1-24h), Days (1-31d), Months (1-12M)
- Days of week restrictions (e.g., weekdays only)
- End conditions (occurrences or timestamp)
- Owner-based access control
- Manual "Run Now" for testing
- Automatic skip of missed executions on system restart

### 3. Web Interface
Complete Angular UI for management:
- **Prompt Manager**: CRUD operations, webhook toggle (admin), URL copy
- **Scheduler Manager**: CRUD operations, enable/disable, run now
- **Search functionality**: Filter by name, description, author
- **Visual indicators**: Owner badges, status badges, webhook status
- **Parameter modes**: Toggle between simple/structured in scheduler form
- **Error feedback**: Backend validation errors displayed in UI

### 4. Backend Architecture
- **PromptExecutionService**: Unified execution logic for all contexts
- **PromptService**: CRUD operations with access control
- **SchedulerService**: Lifecycle management with 30s check interval
- **REST API**: Complete CRUD endpoints for prompts and schedulers
- **Parameter processing**: Automatic detection and conversion
- **Validation**: Strict parameter validation with missing key detection

## Parameter System

The system supports three parameter modes:

### Simple Parameter (`{{PARAMETERS}}`)
```yaml
commands:
  - "Review PR {{PARAMETERS}}"
```
Execution: `executePrompt(promptId, "1234", username)`
Result: `"Review PR 1234"`

### Structured Parameters (`{{key}}`)
```yaml
commands:
  - "Deploy {{app}} to {{env}}"
```
Execution: `executePrompt(promptId, {app: "coday", env: "prod"}, username)`
Result: `"Deploy coday to prod"`

### Append Mode (no placeholders)
```yaml
commands:
  - "Analyze logs"
```
Execution: `executePrompt(promptId, "from last 24h", username)`
Result: `"Analyze logs from last 24h"`

### Prompts
- **Viewing/Editing**: All users (collaborative)
- **Webhook enable/disable**: CODAY_ADMIN only
- Visual indicators show creator with badge

### Schedulers
- **Owner**: Can view/edit/delete own schedulers
- **CODAY_ADMIN**: Can manage all schedulers
- Executions run with creator's identity

### Data Models
```typescript
interface Prompt {
  id: string
  name: string  // normalized: lowercase-with-hyphens
  description: string
  commands: string[]
  webhookEnabled: boolean  // CODAY_ADMIN only
  createdBy: string
  createdAt: string
  updatedAt?: string
}
interface Scheduler {
  id: string
  name: string
  enabled: boolean
  promptId: string
  schedule: IntervalSchedule
  parameters?: Record<string, unknown>
  createdBy: string
  lastRun?: string
  nextRun?: string | null
  occurrenceCount?: number
}
```

### Storage Structure
```
~/.coday/
  projects/
    my-project/
      prompts/
        {prompt-id}.yml
      schedulers/
        {scheduler-id}.yml
```
### API Endpoints
```
# Prompts
GET/POST    /api/projects/:project/prompts
GET/PUT/DELETE  /api/projects/:project/prompts/:id
POST/DELETE     /api/projects/:project/prompts/:id/webhook

# Schedulers
GET/POST    /api/projects/:project/schedulers
GET/PUT/DELETE  /api/projects/:project/schedulers/:id
POST        /api/projects/:project/schedulers/:id/enable
POST        /api/projects/:project/schedulers/:id/disable
POST        /api/projects/:project/schedulers/:id/run

# Webhook execution (unchanged)
POST        /api/webhooks/:promptId/execute
```
### Prompt Manager
- List with search (name, description, author)
- Create/Edit forms with command fields
- Webhook toggle and URL copy (admin only)
- Delete with confirmation
- Owner badges for visibility
### Scheduler Manager
- List with search (name, prompt, author)
- Create/Edit forms with:
  - Prompt selection dropdown
  - Start date/time picker
  - Interval configuration
  - Days of week selector
  - End condition options
  - Parameter mode toggle (simple/structured)
- Enable/Disable toggle
- Run Now button
- Status indicators (enabled/disabled, next run)
### Parameter Processing Logic
```typescript
// In PromptExecutionService.processCommands()
if (typeof parameters === 'string') {
  // Simple mode
  if (hasParametersPlaceholder) {
    // Replace {{PARAMETERS}} in ALL commands
  } else if (hasOtherPlaceholders) {
    // Error: structured placeholders need object
  } else {
    // Append to first command only
  }
} else if (typeof parameters === 'object') {
  // Structured mode: interpolate all {{key}} placeholders
}
// Final validation: check for remaining {{...}}
// Throw error with list of missing keys if any remain
```

### Scheduler Execution Flow
1. SchedulerService checks every 30 seconds
2. Finds schedulers with `nextRun` <= now
3. Converts parameters (detect {PARAMETERS: "value"} → string)
4. Calls PromptExecutionService.executePrompt()
5. Updates lastRun, increments occurrenceCount
6. Calculates nextRun (respecting daysOfWeek, endCondition)
7. Logs execution result

### Error Handling
- Parameter validation errors return HTTP 422 with clear messages
- Missing placeholders: "Missing required parameters: key1, key2"
- Type mismatch: "Prompt contains structured placeholders. Use an object parameter."
- Errors displayed in UI alerts with full backend message

### New Files
- `libs/model/src/lib/prompt.ts` - Prompt data model
- `libs/model/src/lib/scheduler.ts` - Scheduler data model
- `libs/service/src/lib/prompt.service.ts` - Prompt CRUD
- `libs/service/src/lib/scheduler.service.ts` - Scheduler lifecycle
- `libs/service/src/lib/prompt-execution.service.ts` - Unified execution
- `libs/utils/src/lib/interval-schedule.utils.ts` - Interval parsing
- `apps/server/src/lib/prompt.routes.ts` - Prompt REST API
- `apps/server/src/lib/scheduler.routes.ts` - Scheduler REST API
- `apps/server/src/lib/prompt-execution.routes.ts` - Webhook execution
- `apps/client/src/app/components/prompt-manager/` - Prompt UI
- `apps/client/src/app/components/prompt-form/` - Prompt form
- `apps/client/src/app/components/scheduler-manager/` - Scheduler UI
- `apps/client/src/app/components/scheduler-form/` - Scheduler form
- `apps/client/src/app/core/services/prompt-api.service.ts` - Prompt API client
- `apps/client/src/app/core/services/scheduler-api.service.ts` - Scheduler API client
- `docs/04-configuration/prompts.md` - Complete prompt documentation
- `docs/04-configuration/schedulers.md` - Complete scheduler documentation

### Modified Files
- `libs/handler/src/lib/prompt-chain.handler.ts` - Updated for new pattern
- `apps/server/src/server.ts` - Services initialization
- `apps/client/src/app/components/floating-menu/` - Menu integration
- Various service and model updates

### Removed Files
- `docs/04-configuration/webhooks.md` - Replaced by prompts.md
- Legacy webhook CLI handlers (replaced by UI)

## Testing

- ✅ Parameter interpolation (simple, structured, append modes)
- ✅ Validation (missing parameters, type mismatches)
- ✅ Scheduler execution (intervals, daysOfWeek, end conditions)
- ✅ Missed execution handling (skip and recalculate)
- ✅ Access control (owner, CODAY_ADMIN)
- ✅ UI workflows (CRUD, search, filtering)
- ✅ Error display (backend messages in UI)
- ✅ Webhook execution (backward compatible)

## Breaking Changes

⚠️ **Webhook system replaced**: The previous webhook-based approach is now the prompt system. External integrations using `/api/webhooks/:uuid/execute` continue to work but should be updated to use prompt IDs.

⚠️ **Parameter format change**: Webhook templates now support both simple string and structured object parameters.

### For Existing Webhooks
1. Review existing webhook configurations
2. Create equivalent prompts through web UI
3. Enable webhook flag (CODAY_ADMIN)
4. Update external integrations with new prompt IDs
5. Test thoroughly before removing old webhooks

### For Scheduled Tasks
1. Create prompts for your automated tasks
2. Create schedulers with appropriate intervals
3. Set parameters (simple or structured)
4. Enable and test with "Run Now"
5. Monitor execution via logs and UI

## Documentation

Complete documentation added:
- `docs/04-configuration/prompts.md` - Prompt system guide
- `docs/04-configuration/schedulers.md` - Scheduler system guide

Includes:
- Architecture and concepts
- Parameter modes and validation
- Access control details
- Usage examples
- API reference
- Troubleshooting guide
- Best practices

## Related

Closes #411

## Checklist

- [x] Code compiles without errors
- [x] Linter passes
- [x] Parameter system implemented and validated
- [x] Scheduler system implemented and tested
- [x] Web UI complete (prompts + schedulers)
- [x] REST API complete with validation
- [x] Access control implemented
- [x] Error messages propagate to UI
- [x] Documentation complete (prompts.md, schedulers.md)
- [x] Migration guide provided
- [x] Backward compatibility maintained

### CODE REVIEW

File: apps/client/src/app/core/services/trigger-api.service.ts [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+@Injectable({
+  providedIn: 'root',
+})
+export class TriggerApiService {
+  private http = inject(HttpClient)
+
+  /**
+   * Build the base URL for trigger endpoints
+   */
+  private getBaseUrl(projectName: string): string {
```
[the remark this answers is not in the document]
Comment (vincent-audibert-whoz) [reply apps/client/src/app/core/services/trigger-api.service.ts]: Done

**Review by vincent-audibert-whoz — COMMENTED** [review]
This PR introduces a substantial architectural shift from webhooks to a prompts and schedulers system. While the implementation is comprehensive and well-documented, there are critical security issues and architectural concerns that should be addressed.

## 🔴 Critical Security Issues

### 1. Command Injection via Parameter Interpolation
The `processCommands` method in `PromptExecutionService` performs simple string replacement without sanitization:
```typescript
processedCommand = processedCommand.replace(new RegExp(placeholder, 'g'), String(value))
```
**Risk**: If these commands are ever executed in a shell context, this enables arbitrary command injection. Even if not currently executed in shell, this is a dangerous pattern that will likely cause security issues as the codebase evolves.

**Recommendation**: Implement proper escaping/sanitization for all user-provided parameters, or use a safer templating engine with built-in XSS/injection protection.

### 2. Regex Injection Vulnerability
The same line constructs regexes from user-controlled keys:
```typescript
new RegExp(placeholder, 'g')
```
**Risk**: If placeholder contains special regex characters, this can cause regex injection or DoS through catastrophic backtracking.

**Recommendation**: Escape regex special characters or use literal string replacement.

### 3. Missing Rate Limiting on Webhook Endpoint
The `/api/webhooks/:promptId/execute` endpoint has no rate limiting or abuse prevention:
```typescript
app.post('/api/webhooks/:promptId/execute', async (req, res) => {
  // No rate limiting, quota checks, or concurrency limits
  const result = await promptExecutionService.executePrompt(...)
})
```
**Risk**: Denial of service through automated prompt execution spawning unlimited threads and Coday instances.

**Recommendation**: Implement per-user rate limiting, global execution quotas, and maximum concurrent executions per user/project.

### 4. Race Condition in threadLifetime Management
The `resolveThreadId` method has a non-atomic read-check-update pattern:
```typescript
// Check if active thread still exists
const activeThread = await this.threadService!.getThread(projectName, prompt.activeThreadId)
if (!activeThread) {
  const thread = await this.threadService!.createThread(projectName, username, title)
  await this.updatePromptActiveThread(prompt.id, projectName, thread.id) // Race here!
  return thread.id
}
```
**Risk**: Multiple concurrent executions can simultaneously find expired thread, create multiple new threads, and race to update `activeThreadId`, causing inconsistent state and potential data corruption.

**Recommendation**: Implement proper locking (file locks or distributed locks) or use atomic compare-and-swap operations for thread updates.

## ⚠️ Architectural Concerns

### 5. Scheduler Service Memory/Scalability Issues
All schedulers are loaded into memory at startup and kept in a single Map:
```typescript
private readonly schedulers: Map<string, Scheduler> = new Map()

private async loadAllSchedulers(): Promise<void> {
  this.schedulers.clear()
  // Loads ALL schedulers from ALL projects into memory
}
```
**Issue**: This doesn't scale beyond a few dozen projects. The 30-second polling of all schedulers is inefficient.

**Recommendation**: Use event-driven scheduling with proper timer management per scheduler, or use a proper job queue system (Bull, Agenda, etc.).

### 6. Magic Parameter Conversion
The scheduler service has undocumented magic behavior:
```typescript
if (scheduler.parameters && typeof scheduler.parameters === 'object' &&
    Object.keys(scheduler.parameters).length === 1 && 'PARAMETERS' in scheduler.parameters) {
  parameters = String(scheduler.parameters.PARAMETERS) // Magic conversion!
}
```
**Issue**: This implicit behavior isn't documented in types or API, leading to confusion and bugs.

**Recommendation**: Make parameter modes explicit in the type system and validate at API boundaries.

### 7. Violation of Separation of Concerns
`PromptExecutionService` directly manipulates prompt state:
```typescript
private async updatePromptActiveThread(promptId: string, projectName: string, threadId: string) {
  await this.promptService.update(projectName, promptId, { activeThreadId: threadId }, 'system')
}
```
**Issue**: Execution service shouldn't manage prompt storage state. This creates tight coupling.

**Recommendation**: Separate state management into a dedicated service or use proper event-driven architecture.

## 📋 Additional Issues

### 8. Inconsistent Access Control
- Prompts: Anyone can view/edit (collaborative), only CODAY_ADMIN can toggle webhook
- Schedulers: Owner or CODAY_ADMIN can manage
- Webhook execution: Anyone can trigger if enabled

**Issue**: This mixed model is confusing and may not match security requirements.

**Recommendation**: Document the access control model clearly and ensure it aligns with security policy.

### 9. Breaking Changes Insufficiently Addressed
The PR removes the entire webhook system but claims backward compatibility. However:
- Parameter format changes from `{{key}}` to structured objects will break existing integrations
- No automated migration tooling provided
- No clear deprecation timeline

**Recommendation**: Provide migration scripts and maintain webhook compatibility for at least one release cycle.

### 10. Missing Input Validation
Several endpoints lack proper validation:
- Scheduler intervals aren't validated for reasonable bounds (what prevents `1min` with 1 million occurrences?)
- Prompt names lack length limits
- Parameter object depth isn't validated (stack overflow risk)

**Recommendation**: Add comprehensive validation at API boundaries with reasonable limits.

## ✅ Positive Aspects

- Comprehensive documentation in PR description
- Good separation between local and project prompts
- Clear parameter interpolation modes
- Proper logging via CodayLogger
- Well-structured service architecture

## 🎯 Recommended Actions Before Merge

**Must Fix**:
1. Implement input sanitization for parameter interpolation
2. Add rate limiting to webhook endpoints
3. Fix race condition in threadLifetime with proper locking
4. Add validation limits for scheduler intervals and occurrences

**Should Fix**:
5. Redesign scheduler service for scalability (event-driven or job queue)
6. Make parameter modes explicit in type system
7. Separate prompt state management from execution logic
8. Provide migration tooling for existing webhooks

**Nice to Have**:
9. Add integration tests for concurrent execution scenarios
10. Document access control model comprehensively
11. Add parameter injection edge case tests

The scope and ambition of this PR are commendable, but addressing the security issues is critical before production deployment.

File: apps/client/src/app/components/prompt-form/prompt-form.component.html [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        <span class="hint"
+          >Project prompts are committable files and shouldn't have changing thread IDs. Use local prompts for thread
+          reuse.</span
+        >
+      </p>
+    </div>
+    -->
+
+    <!-- Source selection (creation only) -->
+    <div *ngIf="!isEditMode" class="source-section">
```
Comment (charles-monot-whoz) [inline apps/client/src/app/components/prompt-form/prompt-form.component.html]: Agent should be aware of the new template syntax @if
